src/acp/internal/setDefaultAccessControlThingIfNotExist.ts

```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright 2021 Inrupt Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { buildThing, createThing, getIriAll, getSourceUrl } from "../..";
+import { ACP } from "../constants";
+import type { WithAccessibleAcr } from "../acp";
+import { getAccessControlResource } from "./getAccessControlResource";
+import { getAccessControlResourceThing } from "./getAccessControlResourceThing";
+import {
+  DefaultAccessControlName,
+  getDefaultAccessControlUrl,
+} from "./getDefaultAccessControlUrl";
+import { setAccessControlResourceThing } from "./setAccessControlResourceThing";
+
+function getTypeFromDefaultAccessControlName(
+  name: DefaultAccessControlName
+): string {
+  if (name.includes("Member")) {
+    return ACP.memberAccessControl;
+  }
+  return ACP.accessControl;
+}
```

Comment:
I started based on type. This happens to be an internal convenience function based on a literal type. It is unit tested via the functions using it, and I would argue therefore acceptably brittle.

For the sake of simplicity, I haven't made the default access control names configurable. This could be changed at some point, I don't think this need to be more complex than it is for the time being.

Reply:
If `setDefaultAccessControlThingIfNotExist` took a type instead of a name, and `getTypeFromDefaultAccessControlName` became `getNameFromDefaultAccessControlType`, I think the complexity would remain the same, but we'd be using the type to infer the name, and not the other way around, and it sounds more correct to me. I don't think it makes much of a difference in the code, and I agree that unit testing it makes it safe enough, but if we want to discourage our users to do string manipulations on IRIs, and consider them opaque, we should try to constraint ourselves to do the same.

Reply:
Then `setDefaultAccessControlThingIfNotExist` would have to take two arguments: one to say if it's a member policy, one to say if it applies to ACRs.

It is a lot of changes everywhere in the code and it doesn't improve the situation. We have 4 default access controls whose names are constants. It is not string manipulation against IRIs. It is matching on constants.

Reply:
`getDefaultAccessControlThing` is definitively assuming that we can concatenate the name to the ACR IRI, and looks up a thing based on this, that's what I refer to as string manipulation. And if there are four possible names, I would assume there are four associated types, so I don't get why `setDefaultAccessControlThingIfNotExist` would take more arguments than it already does.

Reply:
Because we're dealing with Policies applying to four different contexts:
1. The resource
2. Its access control resource
3. The children of the resource
4. The access control resources of the children of the resource

Reply:
Bear in mind that all this is ducktaping long lived wrong assumptions.

And for sure, as per your other [comment and associated discussion thread](https://github.com/inrupt/solid-client-js/pull/1330#discussion_r752198423), there is something non-ideal maybe in relying on such convention for having 4 default access controls that the SDK works out of.

But the SDK:
1. Doesn't know the notions of Access Control Resource Thing or Access Control Thing
2. Sporadically and inconsistently support external ACP resources.

Considering this, relying on conventional (ah the evil conventions) default access control things to address each of the contexts is not a bad or britlle solution.

Reply:
We actually have had recent conversations about the wrong assumption of unique values associated to a subject/predicate pattern, this is in the same vein.

At least such assumptions are not built into this solution. There is indeed 4 default access controls for each access control resource, say an access control resource `https://example.org/a` would have uniquely identifiable access controls for itself `https://example.org/a#defaultAccessControl`, its ACR `https://example.org/a#defaultAcrAccessControl`, its children `https://example.org/a#defaultMemberAccessControl` and its children's acrs `https://example.org/a#defaultMemberAcrAccessControl`.

It works, it doesn't rely on made up types, it is a better solution ultimately than inventing something that is not understood by authorization servers and it will still be workable once we have proper support for Access Control Resource Things, Access Control Things, all external ACP elements and have cleaned up our APIs and made them consistent.

Reply:
New input to the discussion: having additional types isn't necessarily supported by the authorization server, so the current approach is the only one that could work, regardless if I like it or not ^^